src/v/cluster/tm_stm.h

```diff
@@ -42,6 +43,64 @@ namespace cluster {
 using use_tx_version_with_last_pid_bool
   = ss::bool_class<struct use_tx_version_with_last_pid_tag>;
 
+class tm_stm;
+
+class txlock_unit {
+    tm_stm* _stm;
+    ss::lw_shared_ptr<ssx::semaphore_units> _units;
```

Comment:
it's an implementation detail which doesn't leak behind txlock_unit's boundary. overall txlock_unit is non-copyable so I don't see a problem there

Reply:
ok, I'm curious why we need a shared_ptr<units> instead of units.

Reply:
I find it easier to work with `shared_ptr` and here the unit's restrictions don't help because _units is private and isn't accessible outside of the class (an implementation detail)

Reply:
i still think it shouldn't be a shared_ptr. Units should have a clear ownership. With the shared pointer the ownership is shared in this case ownership of acquired mutex may be shared.

Reply:
I don't see how how it's true in this particular case, but it's 3vs1 so I'm changing it